sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -130,10 +133,22 @@ object DataSourceV2Strategy extends Strategy {
         config)
 
       val filterCondition = postScanFilters.reduceLeftOption(And)
-      val withFilter = filterCondition.map(FilterExec(_, scan)).getOrElse(scan)
+
+      val withFilter = if (filterCondition.exists(hasScalarPythonUDF)) {
+        // add a projection before FilterExec to ensure that the rows are converted to unsafe
+        val filterExpr = filterCondition.get
+        FilterExec(filterExpr, ProjectExec(filterExpr.references.toSeq, scan))
+      } else {
+        filterCondition.map(FilterExec(_, scan)).getOrElse(scan)
+      }
 
       // always add the projection, which will produce unsafe rows required by some operators
-      ProjectExec(project, withFilter) :: Nil
+      if (project.exists(hasScalarPythonUDF)) {
+        val references = project.map(_.references).reduce(_ ++ _).toSeq
+        ProjectExec(project, ProjectExec(references, withFilter)) :: Nil
```

Comment:
Why do we need to add extra Project on top of Filter here?

Reply:
The v2 data sources return `InternalRow`, not `UnsafeRow`. Python UDFs can't handle `InternalRow`, so this is intended to add a projection to convert to unsafe before the projection that contains a python UDF.

Reply:
oh, I see. It is also used to make sure PythonUDF in top Project takes unsafe row input.

Reply:
nit: If we already add Project on top of Filter, we don't need to add another Project here, right?

Reply:
That one was only added if there was a filter and if that filter ran a UDF. This will add an unnecessary project if both the filter and the project have python UDFs, but I thought that was probably okay. I can add a boolean to signal if the filter caused one to be added already if you think it's worth it.

Reply:
Ok. Let's leave as it is now.

Reply:
+1 for leaving as is.